graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -314,6 +291,138 @@ private Timestamp getNowTimestamp() {
+                } finally {
+                    l.unlock();
+                }
+            }
+
+            /**
+             * Edit builder for the node at the given path (creating it and its parents if needed).
+             */
+            public void editBuilder(ExecutionPath path, Consumer<Reports.Trace.Node.Builder> builderConsumer) {
+                Lock l = treeLock.readLock();
+                l.lock();
+                try {
+                    if (isFinalized) {
+                        throw new RuntimeException("Cannot edit builder after protobuf conversion.");
+                    }
+                    Node node = getOrCreateNode(path);
+                    synchronized (node.builder) {
+                        builderConsumer.accept(node.builder);
+                    }
+                } finally {
+                    l.unlock();
+                }
+            }
+
+            /**
+             * Get node for the given path in nodesByPath (creating it and its parents if needed).
+             */
+            @NotNull
+            private Node getOrCreateNode(ExecutionPath path) {
+                // Fast path for when the node already exists.
+                Node current = nodesByPath.get(path);
+                if (current != null) return current;
+
+                // Find the latest ancestor that exists in the map.
+                List<Object> pathSegments = path.toList();
+                int currentSegmentIndex = pathSegments.size();
+                while (current == null) {
+                    if (currentSegmentIndex <= 0) {
+                        // The root path's node is inserted at construction time, so this shouldn't
+                        // happen.
+                        throw new RuntimeException("root path missing from nodesByPath?");
+                    }
+                    currentSegmentIndex--;
+                    ExecutionPath currentPath = ExecutionPath.fromList(pathSegments.subList(0, currentSegmentIndex));
+                    current = nodesByPath.get(currentPath);
+                }
+
+                // Travel back down to the requested node, creating child nodes along the way as
+                // needed.
+                for (; currentSegmentIndex < pathSegments.size(); currentSegmentIndex++) {
+                    Node parent = current;
+                    ExecutionPath childPath = ExecutionPath.fromList(pathSegments.subList(0, currentSegmentIndex + 1));
+                    Object childSegment = pathSegments.get(currentSegmentIndex);
+
+                    // Note that the correctness of this statement depends on ConcurrentHashMap's
+                    // guarantee that computeIfAbsent will execute the lambda at most once.
+                    current = nodesByPath.computeIfAbsent(childPath, __ -> {
+                        Reports.Trace.Node.Builder childBuilder = Reports.Trace.Node.newBuilder();
+                        if (childSegment instanceof Integer) {
```

Comment:
The old version actually requires this to be true iff this isn't the final path element. Is that assertion reasonable to maintain?

Reply:
It's reasonable to maintain the assertion, I'll add it back in.